README.md
# Gilded Rose

## The Problem
Hi and welcome to team Gilded Rose. As you know, we are a small inn with a prime location in a prominent city run by a 
friendly innkeeper named Allison. We also buy and sell only the finest goods. 
Unfortunately, our goods are constantly degrading in quality as they approach their sell by date. 
We need you to write a system that allows us to manage our inventory, 
so that we are able to service all of the adventurers who frequent our store 
(we don't want to run out of healing potions when an tiefling comes in unlike last time - poor Leeroy).

Here are the basic rules for the system that we need:

1. All items have a SellIn value which denotes the number of days we have to sell the item
2. All items have a Quality value which denotes how valuable the item is
3. At the end of each day our system lowers both values for every item

Since this is the real world, there are some edge cases we need for you to account for as well:

1. Once the sell by date has passed, Quality degrades twice as fast
2. The Quality of an item is never negative
3. "Aged Brie" actually increases in Quality the older it gets
4. The Quality of an item is never more than 50
5. "Sulfuras", being a legendary item, never has to be sold or decreases in Quality
6. "Backstage passes", increases in Quality as it's SellIn value approaches; 
  Quality increases by 2 when there are 10 days or less and by 3 when there are 5 days or less but Quality drops to 0 after the concert
7. "Conjured" items degrade in Quality twice as fast as normal items
8. An item can never have its Quality increase above 50, however "Sulfuras" is a legendary item and as such its Quality is 80 and it never alters.

We currently keep our inventory in a hand written list. Since Allison wants to get home at night, we keep the writing to a minimum. 
Each line has the following information, in order:

1. Item Name
2. Item Category
3. Sell In
4. Quality

### Additional Requirements:
1. There is no requirement for what you choose as your interface into the system, however whatever interface you choose should, 
at a minimum, provide for the following commands:
	1. Ask for the entire list of inventory
	2. Ask for the details of a single item by name
	3. Progress to the next day
	4. List of trash we should throw away (Quality = 0)
2. In this repo, you will find an inventory.txt file. This is the initial inventory your solution should load. 
After that, you may store the data however you wish.

## The Fine Print
Please use whatever technology and techniques you feel are applicable to solve the problem. 
We suggest that you approach this exercise as if this code was part of a larger system. 
The end result should be representative of your abilities and style.

Please fork this repository, then when you have completed your solution, 
issue a pull request to notify us that you are ready for us to review your submission.

Have fun.

## Things To Consider
Here are a couple of thoughts about the domain that could influence your response:

* The world is a magical place - you never know when the next "special requirement" might pop up - how can you make this painless?
* Keep in mind that accurate inventory is a must for the shop, 
how might you ensure that the future programmer who takes over the code while you are off adventuring doesn't mistakenly mess things up?

CHANGES.txt
Problem assembling:  In this version update it was decided to achieve the following goals:
1. Entirely update code to ES6 & ES7

2. Defining the feature purpose and UX for this project.
:: Constraints - 
:: C1. The focus this iteration is to be able to explain design choices under time constraints of two days
:: C2. A single webpage supported by REST API backend.
:: C3. Maintain using a file for data manipulation and not a DB to ensure portability and stability by not adopting additional configuration
:: C4. Re-utilize inventory data as modularization is already supported.
:: C5. Maintain original terminal command line controls (modularization already supported)
:: C6. Maintain the original Guilded Rose Inventory reqs
:: Requirements - 
:: R1. Show all the items.
:: R2. Increment the days of use.
:: R3. Purchase an item.
:: R4. Sell an item.
:: R5. Reset the item list.
:: R6. Random Gossip.

2. Create a webpage using Angular

3. Create a listening server with REST API routes to following CRUD to support the UX reqs.
:: Use object pointers (shallow copy) to inventory data to facilate data manipulation.
:: GET  request - Show all the items.
:: POST request - Increment the days of use.
:: POST request -  Purchase an item.
:: DELETE request - Sell an item.
:: POST request - Reset the item list.
:: GET request - Get Gossip

Conclusion: Reflections of choices and recommended enhancements
1.  Define Story ID for the UX requirements
2.  Create standard message structure of REST API to handle success & error responses.
3.  Enable internal error codes to depict where in the UX story the error is reported by using the Story IDs as a prefix to the error code.
4.  Time to design a more verbose UI to depict a better UX.
5.  Authorization of endpoint usage.
6.  Inventory-api contains gossip endpoint.  It should really be in a seperate api file that depicts the purpose of gossip chatting endpoints.
7.  Not all area's of code is setup for ES6 & ES7.  Use IDE to setup testing to ensure the standard.
8.  Enable Trash feature
9.  Enable front-end UI to be the prospective the Patron talking to the Item Keeper.
10. Enable feature  'Stop going to the store(end program)' in UI to store the current operational inventory


guildedRoseApp.js
"use strict";

// Add in required 3rd party libraries
let express = require('express')
let bodyParser = require('body-parser')
let path = require('path');
const logger = require('morgan');
let readline = require('readline');

// Add in required product js libraries
let ChatResponse = require('./includes/ChatResponse.js');
let InventoryManager = require('./includes/InventoryManager.js');

// Pull in the inventory at startup
InventoryManager.uploadInventory();

let promtpOutput = "";
let patronbuy = false;
let patronsell = false;
let showitem = false;

// Begin to setup web server
let app = express();

// Set webserver configuration
global.appRoot = path.resolve(__dirname);
app.set('views', __dirname + '/views');
app.engine('html', require('ejs').renderFile);
app.set('view engine', 'html');

app.use(logger('dev'));
// Handle intercept OPTIONS method
app.use(function(req, res, next) {
    if (req.method === 'OPTIONS') {
        res.header('Access-Control-Allow-Origin', '*');
        res.header('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE,OPTIONS');
        res.header('Access-Control-Allow-Headers', 'Content-Type, Authorization, Content-Length, X-Requested-With');
        res.send(200);
    }
    else {
        next();
    }
});
app.use(bodyParser.json({ limit: '5mb' }));
app.use(express.cookieParser());
app.use(express.static(__dirname + '/www'));
app.use(app.router);

app.use(function (error, req, res, next) {
    if (!error) {
        next();
    } else {
        console.error('My stack error:  ', error.stack);
        res.send(500);
    }
});

// Define REST API
const inventoryApi = require('./routes/inventory-api');

// Link data connectors to REST API routes
inventoryApi.linkInventoryData(InventoryManager);
inventoryApi.linkCharResponder(ChatResponse);

// Initialize REST API routes to the webserver
inventoryApi.router(app);

app.get('/test', function(req, res){
   res.end('got a connection');
});

// Set and initalize only one webpage
app.get('/', function(req, res, next){
    res.status(200).render('inventory-main');
});

module.exports = app;

// Turn on webserver to listen on desired port
if (require.main === module) {
  app.listen(4000, () => console.log('Listening in dev mode on port 4000'));
  app.on('error', e => console.error("Webserver error:  ",e));
}

// ================================================
// Create a menu prompt for commandline interaction
let rl = readline.createInterface(process.stdin, process.stdout);

promtpOutput = ChatResponse.getIntroduction() + '\n' + ChatResponse.displayMenuOptions();
rl.setPrompt(promtpOutput);
rl.prompt();

//const menu = [ "0. Tell me some local gossip",
//   "1. Ask for the entire list of inventory",
//	"2. Ask for the details of a single item by name",
//	"3. Progress to the next day",
//	"4. List of trash we should throw away (Quality = 0)",
//	"5. Buy",
//	"6. Sell",
//	"7. Restart inventory",
//   "8. Never Come Back"]
   
rl.on('line', function(line) {
   if (line === "0" ) { 
      // Get local drama
      promtpOutput = ChatResponse.getRandomGossip() + '\n' + '\n' + ChatResponse.displayMenuOptions();
   }
   else if ( line === "1" ) {
      // 1. Ask for the entire list of inventory
      promtpOutput = InventoryManager.showInventory() + '\n' + '\n' + ChatResponse.displayMenuOptions();
   }
   else if ( line === "2" ) {
      // 2. Ask for the details of a single item by name
	  showitem = true;
	  promtpOutput = ChatResponse.displayShowInfo();
   }
   else if ( line === "3" ) {
      // 3. Progress to the next day
      InventoryManager.incrementDay();
      promtpOutput = ChatResponse.displayMenuOptions();
   }
   else if ( line === "4" ) {
      // 4. List of trash we should throw away (Quality = 0)
      promtpOutput = InventoryManager.showTrashInventory() + '\n' + '\n' +  ChatResponse.displayMenuOptions();
   }
   else if ( line === "5" ) {
      // 5. Patron Buy"
      patronbuy = true;
      promtpOutput = ChatResponse.displayStoreSellsInfo();
   }
   else if ( line === "6" ) {
      // 6 Patron Sell
      patronsell = true;
      promtpOutput = ChatResponse.displayStoreBuysInfo();
   }
   else if ( line === "7" ) {
      // 7. Restart program
      InventoryManager.restartInventory();
      promtpOutput = InventoryManager.showInventory() + '\n' + ChatResponse.displayMenuOptions();
   }
   else if ( line === "8" ) {
      // 8. Never Come Back
      // Store inventory for next time
      InventoryManager.packUpInventory();
      rl.close();
   }
   else if (showitem) {
      promtpOutput = InventoryManager.showItem(line) + '\n' + '\n' +  ChatResponse.displayMenuOptions();
      showitem =  false;
   }   
   else if (patronbuy) {
      if (InventoryManager.removeItem(line)) {
         promtpOutput = ChatResponse.patronBoughtDisplayMenuOptions(line);
      }
      else {
         promtpOutput = ChatResponse.errorDisplayMenuOptions();
      }
      patronbuy =  false;
   }
   else if (patronsell) {
      if (InventoryManager.addItem(line)) {
         promtpOutput = ChatResponse.patronSoldDisplayMenuOptions(line) ;
       }
      else {
         promtpOutput = ChatResponse.errorDisplayMenuOptions();
      }
      patronsell =  false;
   }
   else {
      promtpOutput = ChatResponse.warningDisplayMenuOptions();
   }

   rl.setPrompt(promtpOutput);
   rl.prompt();

}).on('close',function(){
    process.exit(0);
});




routes/inventory-api.js

"use strict";

let localInventoryManager = {};
let localChatResponse = {};

// Forward data linkage pointer into local assignment for routes to manipulate
// Only a shallow copy 
const linkInventoryData = function (InventoryManager){ 
    localInventoryManager = InventoryManager; 
    console.log('TEST localInventoryManager Output:', localInventoryManager.showInventory());
};

// Forward function linkage pointer into local assignment for routes to utilize
// Only a shallow copy 
const linkCharResponder = function (ChatResponse){ 
    localChatResponse = ChatResponse; 
    console.log('TEST localChatResponse Output:', localChatResponse.getRandomGossip());
};

const apiHeaders = {'Content-Type':'application/json', "Access-Control-Allow-Origin": "*"};

const errorHandling = (err, res)=>{
    console.log("errorHandling res.headers:  ", res.headers);
        console.log("errorHandling res.body:  ", res.body);
    if (err.responseCode){
        res.status('errorHandling:  ',err.responseCode);
    } else {
        console.log('errorHandling:  ', err);
        res.status(500);
    }
    res.set(apiHeaders).send({message: err.message});
};


const router = function(app){

    // GET  request - Show all the items.
    app.get('/api/v1/inventory/list', (req, res, next)=>{
        // Data is not a promise as it is already stored locally.
        // No .then((data)=> {}) is necessary
        res.status(200).set(apiHeaders).send(localInventoryManager.showInventory());
    });

    // POST request - Increment the days of use.
    app.post('/api/v1/inventory/increment', (req, res, next)=>{
        // Data is not a promise as it is already stored locally.
        // No .then((data)=> {}) is necessary
        localInventoryManager.incrementDay();
        res.status(200).set(apiHeaders).send();
    });

    // POST request -  Purchase an item.
    app.post('/api/v1/inventory/purchase', (req, res, next)=>{
        // Data is not a promise as it is already stored locally.
        // No .then((data)=> {}) is necessary
        try {
            localInventoryManager.addItemAsObject(req.body);
            res.status(200).set(apiHeaders).send();
        }
        catch (err) {
            errorHandling(err, res);
        }
    });

    // DELETE request - Sell an item.
    app.delete('/api/v1/inventory/:itemIndex/sell', (req, res, next)=>{
        // Data is not a promise as it is already stored locally.
        // No .then((data)=> {}) is necessary
        try {
            console.log('req.params=', req.params);
            const itemIndex = parseInt(req.params.itemIndex);
            console.log('itemIndex=', itemIndex);
            localInventoryManager.removeItemAsIndex(itemIndex);
            res.status(200).set(apiHeaders).send();
        }
        catch (err) {
            errorHandling(err, res);
        }

    });

    // POST request - Reset the item list.
    app.post('/api/v1/inventory/reset', (req, res, next)=>{
        // Data is not a promise as it is already stored locally.
        // No .then((data)=> {}) is necessary
        localInventoryManager.restartInventory();
        res.status(200).set(apiHeaders).send();
    });   

    // GET request - Get Gossip
    app.get('/api/v1/inventory/gossip', (req, res, next)=>{
        // Data is not a promise as it is already stored locally.
        // No .then((data)=> {}) is necessary
        const gossip = localChatResponse.getRandomGossip();
        //console.log('getGossip - loading done gossip=', gossip);
        res.status(200).set(apiHeaders).send({gossip});
    });
};

// Provide public access to the desired functions
module.exports = {
    linkInventoryData : linkInventoryData,
    linkCharResponder : linkCharResponder,
    router: router
};


includes/ChatResponse.js
// Allison's verbal responses

const introduction = `Hi and welcome to the Gilded Rose. Welcome to our small inn within our prominent city.
   I am Allison. We buy and sell only the finest goods.`;

// 15 gossip drama sayings 
const drama = [ "Beware of Leeroy Jenkins.  He isn't a team player",
   "Orcs seem to buy the most potions.",
   "I heard that a dragon stole a dwarvens king gold and kingdom.  And sits on top of a huge pile of dwarvens gold, valuables and relics.",
   "Did you know that the king came by my shop last month and said I was sweet?  I sure hope he comes by again soon.",
   "I heard the king pulled his sword from a stone",
   "The best healing potion to buy is Gothrog's flesh rejuvenation, you will see the hole in you regenerate by placing it around the wound",
   "The bad thing about Gothrog's flesh rejuvenation potion where to you get these ingredients: Phoenix fire, elder vampire blood, Jelly from Gothrog's Ooze Monster, Ender Apple Cider vinegar", 
   "Serious Gothrog where and how do you get a Phoenix, elder vampires are over 1000 years old, and you literally created your own ooze monster",
   "I heard to create Gothrog's ooze monster he pulls the essence of life from a gaint acidic toad tadpole and puts it in an oozing soul stone",
   "I heard that Gothrog's ooze monsters are spell summoned into battles",
   "I heard the dark shogans and ether demegorgons are hunting Gothrog for his inifinty kane spear, the spear is a jeweled relic from the civilation before the demon wars",
   "The only reason Gothrog's healing potions are even around is because he left 100s chests full of them on the during the last battle of the Rose War",
   "The different kingdoms, guilds, mercs and heros took Gothrog's health potion chests during the last battle of the Rose War",
   "No one has seen Gothrog since the last Rose battle.  Probably better that way as he is cheered by many, but hated by just as much.",
   "I'm only a 1377 years old elf, a little flightly and I care to gossip about what I want",
   "Sulfuras was assassinated by the dark shogans through a trap planned out by the demegorgons"   ]

const instructions = `Type the number of the following option`;

const menu = [ "0. Tell me some local gossip",
   "1. Ask for the entire list of inventory",
	"2. Ask for the details of a single item by name",
	"3. Progress to the next day",
	"4. List of trash we should throw away (Quality = 0)",
	"5. Patron Buys",
	"6. Patron Sells",
    "7. Restart inventory (From orginal)",
    "8. Stop going to the store(end program)" ]

const seeYouTomorrow = `I'll see you tomorrow'\n''\n'`; 

function randomIntFromInterval(min,max) { // min and max included
    return Math.floor(Math.random()*(max-min+1)+min);
}

exports.getIntroduction = function()  {
   return (introduction);
}

exports.getRandomGossip = function()  {
   const randomSelection = randomIntFromInterval(0,drama.length-1);
   return (drama[randomSelection]);
}

exports.displayMenuOptions = function()  {
   let retVal = instructions + '\n';
   for ( let index = 0; index < menu.length; index++ ) {
      retVal =  retVal + menu[index] + '\n';
   }
   retVal =  retVal + '\n';
   return (retVal);
}

exports.displayShowInfo = function()  {
   const retVal = `Enter in the exact item name to see:  `;
   return (retVal);
}

exports.displayStoreBuysInfo = function()  {
   const retVal = `In csv form enter the item name, category, selling time (integer), quality(integer)
      (example: Roast Rat Burgers,Food,7,14):  `;
   return (retVal);
}

exports.displayStoreSellsInfo = function()  {
   const retVal = `Enter in the exact item name to purchase:  `;
   return (retVal);
}

exports.patronBoughtDisplayMenuOptions = function(line)  {
   let retVal = `Patron, you bought the item:  ${line}'\n''\n'instructions'\n'`;
   for ( let index = 0; index < menu.length; index++ ) {
      retVal =  `${retVal}${menu[index]}'\n'`;
   }
   retVal = retVal + seeYouTomorrow;
   return (retVal);
}

exports.patronSoldDisplayMenuOptions = function(line)  {
   let retVal = `Patron, you sold the item:  ${line}'\n''\n'${instructions}'\n'`;
   for ( let index = 0; index < menu.length; index++ ) {
      retVal =  `${retVal}${menu[index]}'\n'`;
   }
   retVal = retVal + seeYouTomorrow;
   return (retVal);
}


exports.warningDisplayMenuOptions = function()  {
   let retVal = `You didn't select a menu item.'\n''\n'${instructions}'\n'`;
   for ( let index = 0; index < menu.length; index++ ) {
      retVal =  `${retVal}${menu[index]}'\n'`;
   }
   return (retVal);
}

exports.errorDisplayMenuOptions = function()  {
   let retVal = `That didn't work try again.'\n''\n'${instructions}'\n'`;
   for ( let index = 0; index < menu.length; index++ ) {
      retVal =  `${retVal}${menu[index]}'\n'`;
   }
   return (retVal);
}


www/scripts/inventory-main.js
let adminApp = angular.module('inventoryMain', ['ckeditor']);
adminApp.config(['$httpProvider', function($httpProvider) {
    //initialize get if not there
    if (!$httpProvider.defaults.headers.get) {
        $httpProvider.defaults.headers.get = {};
    }

    // Answer edited to include suggestions from comments
    // because previous version of code introduced browser-related errors

    //disable IE ajax request caching
    $httpProvider.defaults.headers.get['If-Modified-Since'] = 'Mon, 26 Jul 1997 05:00:00 GMT';
    // extra
    $httpProvider.defaults.headers.get['Cache-Control'] = 'no-cache';
    $httpProvider.defaults.headers.get['Pragma'] = 'no-cache';

    $httpProvider.defaults.headers.common["X-Requested-With"] = 'XMLHttpRequest';
}]);

adminApp.controller('mainCtrl', ['$scope', '$http', '$q', '$timeout', '$animate', '$interval', function($scope, $http, $q, $timeout, $animate, $interval){
    let m = $scope.model = {
        itemData: {  "ItemName" : "", "ItemCategory" : "", "Sellin" : "", "Quality" : 0, "Sell Item" : 0 } 
    };

    let init = function(){
        //console.log('---------------In the INIT of mainCtrl-------------');

        // No auth. Go right in.
        m.loading = true;
        m.gossip = null;
        m.inventoryData = null;
        angular.copy(m.itemData, m.anItem);  // Deep Copy

        // Get inventory
        return $q.all([
            $http.get(`/api/v1/inventory/list`),
           // $http.get(`/api/v1/gossip/all`)
        ]).then(function(response){
            m.inventoryData = response[0].data;
            //console.log('m.inventoryData=', m.inventoryData);
            
            // Future features for gossip.
            //m.gossipData = values[1].data;
            m.loading = false;

        }).catch(function(response){
            swal({
                title: "Something went wrong.",
                text: "Sorry, the service is unavailable right now.",
                icon: "error",
                button: "Okay",
            });
            m.loading = false;
            throw response.data ? new Error(response.data.message) : response;
        });
    };

    $scope.getItemList = function(){
        m.loading = true;
        return $http.get(`/api/v1/inventory/list`).then(function(response){
            console.log('get item list - loading done');
            m.inventoryData = response.data;
            m.loading = false;
        });
    };

    $scope.incrementInventory = function(){
        m.loading = true;
        return $http.post(`/api/v1/inventory/increment`).then(function(response){
            console.log('incrementInventory()');
            $scope.getItemList();
        });
    };

    $scope.purchaseItem = function(itemObject){
        m.loading = true;
        //console.log('itemObject=',itemObject);
        return $http.post(`/api/v1/inventory/purchase`,itemObject).then(function(response){
            console.log('purchaseItem()');
            $scope.getItemList();
        });
    };

    $scope.sellItem = function(itemIndex){
        m.loading = true;
        //console.log('itemIndex=',itemIndex);

        return $http.delete(`/api/v1/inventory/${itemIndex}/sell`).then(function(response){
            console.log('sellItem()');
            $scope.getItemList();
        });
    };

    $scope.resetInventory = function(){
        m.loading = true;
        return $http.post(`/api/v1/inventory/reset`).then(function(response){
            console.log('resetInventory()');
            $scope.getItemList();
        });
    };

    $scope.getGossip = function(){
        m.loading = true;
        return $http.get(`/api/v1/inventory/gossip`).then(function(response){
            console.log('getGossip - loading done response.data=', response.data);
            m.gossip = response.data.gossip;
            m.loading = false;
        });
    };

    $scope.$watch(('model.inventoryData'), function(newVal, oldVal){
        if (newVal && newVal !== oldVal){

            // For future use
            //console.log("$watch(('model.inventoryData') - with NO auth");
            //console.log("$watch(('model.inventoryData') -newVal=", newVal);
            //console.log("$watch(('model.inventoryData') -oldVal=", oldVal);
        }
    }, true)

    init();

}]);



includes/InventoryManager.js
const csvjson = require('csvjson');
const fs = require('fs');

const inventoryfileorg = "inventory-org.csv";
const inventoryfile = "inventory.csv";
const csvoptions = {
	headers : "ItemName,ItemCategory,Sellin,Quality",
    delimiter : ',' , // optional
    quote     : '"' // optional
};

/*
1. Item Name
2. Item Category
3. Sell In
4. Quality
*/
const templateItemObject = { "ItemName" : "", "ItemCategory" : "", "Sellin" : 0, "Quality" : 0 };


// Array of objects
let inventory = [];


// Loads the inventory at startup
exports.uploadInventory = function()  {
   // Convert CSV to JSON object
   const inventory_file_data = fs.readFileSync(inventoryfile, { encoding : 'utf8'});
   inventory = csvjson.toObject(inventory_file_data, csvoptions);
   //console.log("inventory=", inventory);
}

// Shows the inventory
exports.showInventory = function()  {
   let strInventory = JSON.stringify(inventory, null, 2);
   //console.log("strInventory=", strInventory);
   strInventory = strInventory + '\n';
   return (strInventory);
}

// Shows the inventory
exports.showItem = function(line)  {
	let retVal = "I couldn't find the item:  " + line;
	let stopFlag = false;
	for ( let inc = 0; inc < inventory.length && !stopFlag; inc++ ) {
		//console.log("inventory[inc].ItemName=", inventory[inc].ItemName);
		if ( inventory[inc].ItemName === line ) {
			retVal = JSON.stringify(inventory[inc], null, 2) + '\n';
			stopFlag = true;
		}
	}
   return (retVal);
}

// Increment day and affect inventory
//1. All items have a SellIn value which denotes the number of days we have to sell the item
//2. All items have a Quality value which denotes how valuable the item is
//3. At the end of each day our system lowers both values for every item
// ADDITIONAL
//1. Once the sell by date has passed, Quality degrades twice as fast
//2. The Quality of an item is never negative
//3. "Aged Brie" actually increases in Quality the older it gets
//4. The Quality of an item is never more than 50
//5. "Sulfuras", being a legendary item, never has to be sold or decreases in Quality
//6. "Backstage passes", increases in Quality as it's SellIn value approaches; 
//  Quality increases by 2 when there are 10 days or less and by 3 when there are 5 days or less but Quality drops to 0 after the concert
//7. "Conjured" items degrade in Quality twice as fast as normal items
//8. An item can never have its Quality increase above 50, however "Sulfuras" is a legendary item and as such its Quality is 80 and it never alters.
exports.incrementDay = function()  {
   for ( let index = 0; index < inventory.length; index++ ) {
      if ( "Aged Brie" === inventory[index].ItemName && inventory[index].Quality < 50 && inventory[index].Sellin > 0 ) {
         inventory[index].Quality++;
      }
      else if ( "Sulfuras" === inventory[index].ItemCategory || "Gothrog" === inventory[index].ItemCategory ) {
         inventory[index].Quality = 80;
      }
      else if ( "Backstage Passes" === inventory[index].ItemCategory ) {

         if ( inventory[index].Sellin === 0 ) {
            inventory[index].Quality = 0;
         }
         else if ( inventory[index].Quality < 50 ) {
            if ( inventory[index].Sellin <= 5 && 
               inventory[index].Quality < 48 ) {
               inventory[index].Quality++;
               inventory[index].Quality++;
            }
            else if ( inventory[index].Sellin <= 10 && 
               inventory[index].Quality < 49 ) {
               inventory[index].Quality++;
            }
            inventory[index].Quality++;
         }
      }
      else if ( "Conjured" === inventory[index].ItemCategory ) {
         inventory[index].Quality = inventory[index].Quality  - 2;
      }
      else {
         if (inventory[index].Quality > 0 && inventory[index].Sellin === 0 ) {
            inventory[index].Quality = inventory[index].Quality - 2;
         }
         else if (inventory[index].Quality > 0 ) {
            inventory[index].Quality--;
         }
		 
      }
   
      if (inventory[index].Sellin > 0 && 
	     !( "Sulfuras" === inventory[index].ItemCategory || "Gothrog" === inventory[index].ItemCategory)) {
         inventory[index].Sellin--;
      }

   }
   const strInventory = JSON.stringify(inventory, null, 2);
   //console.log("strInventory=", strInventory);
   return (strInventory);
}

// Add Item to the inventory
exports.addItem = function(line)  {
   let retVal = false;
   try {
      const purchasedItem = csvjson.toObject(line, csvoptions); 
	  //console.log("purchasedItem=", purchasedItem);

      // Enable a deep copy
      const strTemplateInventory = JSON.stringify(templateItemObject, null, 2);
      let templateInventory = JSON.parse(strTemplateInventory);
      templateInventory.ItemName = purchasedItem[0].ItemName;
      templateInventory.ItemCategory = purchasedItem[0].ItemCategory;
      templateInventory.Sellin = parseInt(purchasedItem[0].Sellin);
      templateInventory.Quality = parseInt(purchasedItem[0].Quality);
      //console.log("templateInventory=", templateInventory);

      inventory[inventory.length] = templateInventory;
      retVal = true;
   }
   catch (err) {
      console.log("Item most likely not added - Error:", err);   
      console.log("Error: line input not a CSV format=", line);   
   }
   return (retVal);
}

// Add Item to the inventory when parameter is an object
exports.addItemAsObject = function(itemObj)  {
   let retVal = false;
   try {
      // Enable a deep copy
      const strTemplateInventory = JSON.stringify(itemObj, null, 2);
      let templateInventory = JSON.parse(strTemplateInventory);
      inventory[inventory.length] = templateInventory;
      retVal = true;
   }
   catch (err) {
      console.error("Item most likely not added - Error:", err);   
      console.error("Error: itemObj is not in the right format=", itemObj);   
   }
   return (retVal);
}

// Remove Item from the inventory by name
exports.removeItem = function(line)  {
   let retVal = false;
   for ( let index = 0; index < inventory.length && !retVal; index++ ) {
      if ( line === inventory[index].ItemName ) {
         inventory.splice(index, 1);
         retVal = true;
      }
   }
   return (retVal);
}

// Remove Item from the inventory by itemIndex
exports.removeItemAsIndex = function(itemIndex)  {
   let retVal = false;
   if ( -1 < itemIndex && itemIndex < inventory.length ) {
      inventory.splice(itemIndex, 1);
      retVal = true;
   }
   return (retVal);
}

// Shows the bad items in inventory
exports.showTrashInventory = function()  {
   // Compute a bad list.
   let trashItems = [];
   let incTrack = 0;
   for ( let inc = 0; inc < inventory.length; inc++ ) {
      if ( inventory[inc].Quality === 0 ) {
         trashItems[incTrack] = inventory[inc];
         incTrack++;
      }
   }
   const strInventory = JSON.stringify(trashItems, null, 2);
   //console.log("strInventory=", strInventory);
   return (strInventory);
}

// Restart the inventory at startup
exports.restartInventory = function()  {
   // Convert CSV to JSON object
   const inventory_file_data = fs.readFileSync(inventoryfileorg, { encoding : 'utf8'});
   inventory = inventory.splice(0,inventory.length); // Remove all elements in the array
   inventory = csvjson.toObject(inventory_file_data, csvoptions);
   //console.log("inventory restarted=", inventory);
}

// Loads the inventory at startup
exports.packUpInventory = function()  {
   // Convert JSON to CSV string
   //"ItemName" : "", "ItemCategory" : "", "Sellin" : 0, "Quality" : 0
   let inventory_file_data = "";
   for (let index = 0; index < inventory.length; index++ ) {
	   inventory_file_data = inventory_file_data + inventory[index].ItemName + "," + 
	      inventory[index].ItemCategory + "," + inventory[index].Sellin + "," +  inventory[index].Quality + '\n'; 
   }
   console.log("inventory_file_data=", inventory_file_data);

   fs.writeFileSync(inventoryfile, inventory_file_data, 'utf8', function(err) {
       if(err) {
           return console.error("packUpInventory failed to write, error:  " ,err);
       }
       console.log("The " + inventoryfile + "file was saved!");
   }); 
}


